module-info-runtime/src/main/java/io/micronaut/module/info/runtime/MicronautModuleRuntimeInfoFactory.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2017-2021 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.module.info.runtime;
+
+import io.micronaut.context.ApplicationContext;
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.module.info.AbstractMicronautModuleInfo;
+import io.micronaut.module.info.MicronautModuleInfo;
+import io.micronaut.module.info.MicronautModuleInfoLoader;
+import jakarta.inject.Singleton;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.Map;
+import java.util.HashMap;
+
+/**
+ * Factory that creates runtime bean definitions for {@link MicronautRuntimeModule} instances.
+ * This factory loads all {@link MicronautModuleInfo} instances and converts them to
+ * resolved {@link MicronautRuntimeModule} instances with proper parent/child relationships.
+ */
+@Internal
+@Factory
+class MicronautModuleRuntimeInfoFactory {
+
+    @Singleton
+    public List<MicronautRuntimeModule> getAllModules(ApplicationContext context) {
```

Comment:
you could just define `List<MicronautModuleInfo> moduleInfos` as a method argument